### ISSUE TITLE
Add convenience constructors for SkyCube

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#905]
@leajouvin is presenting the cube tutorial, where she introduced a helper function to create an empty SkyCube, see [here](http://nbviewer.jupyter.org/github/gammapy/gammapy-extra/blob/master/notebooks/cube_analysis_part1.ipynb#Define-a-method-to-create-an-empty-SkyCube).

This was confusing to some people, Gammapy itself should make it easy to create empty sky cubes in common ways. I think we should add one or two classmethods to `SkyCube` to help with that.

Here's what we have now:
* [gammapy.cube.SkyCube](http://docs.gammapy.org/en/latest/api/gammapy.cube.SkyCube.html)
* [gammapy.cube.SkyCube.empty](http://docs.gammapy.org/en/latest/api/gammapy.cube.SkyCube.html#gammapy.cube.SkyCube.empty)
* [gammapy.cube.SkyCube.empty_like](http://docs.gammapy.org/en/latest/api/gammapy.cube.SkyCube.html#gammapy.cube.SkyCube.empty_like)

We do have separate classes / convenience constructors to set up sky images and energy grids already:
* [gammapy.image.SkyImage](http://docs.gammapy.org/en/latest/api/gammapy.image.SkyImage.html)
* [gammapy.utils.energy.Energy](http://docs.gammapy.org/en/latest/api/gammapy.utils.energy.Energy.html)
* [gammapy.utils.energy.EnergyBounds](http://docs.gammapy.org/en/latest/api/gammapy.utils.energy.EnergyBounds.html)

Since it's orthogonal, maybe we should just say that to define an empty cube one always makes an empty sky image, and empty energy or energy bounds object, and then the SkyCube, i.e. like this:
```
image = SkyImage<some method and parameters>
energy = Energy<some method and parameters>
cube = SkyCube.empty(image, energy)
```
or if you want to avoid the local variables of `image` and `energy`, you could do this:
```
cube = SkyCube.empty(
    image=SkyImage<some method and parameters>
    energy=Energy<some method and parameters>
)
```

@kosack @JouvinLea @adonath @registerrier  - Thoughts on what API we should put?

Anyone has time to implement something here in the coming days?

**Comment by adonath** [comment]
I think we should definitely add a method `SkyCube.empty_like_skyimage()` or similar, that allows creating a cube from already **existing** sky images. This is currently not possible.

In general I don't think we gain much by separating the construction in an energy and image part beforehand. It's just writing even more code. It's a fact that to create an empty `SkyCube` you have to specify O(10) input parameters to define the WCS and energy grid (it's the same e.g. with `gtbin`). This can't be avoided. Yes, one could split the creation into multiple steps by creating the energy and image grid separately. But does that really help?

What I did for my analyses was to define a default set of parameters as a dict `DEFAULT_CUBE` in a central place, so that I can create new SkyCube objects with one line of code:
```
cube = SkyCube.empty(**DEFAULT_CUBE)
```
I don't think it can get much easier than that.

**Comment by adonath** [comment]
Some further thoughts:

I would keep the `.empty()` method as is. I.e. it just takes a set of float/string parameters to create an empty cube. This is equivalent to `SkyImage.empty()`.
In addition we could add `SkyCube.empty_like_skyimage(image, energy)`, which allow creation with a `SkyImage` and `Energy` or `EnergyBounds` object.

**Comment by registerrier** [comment]
Why not have all possible inputs among the list of arguments and test to know which solution to apply?

This would be simpler to have a single empty constructor, than having empty_likeXXX, empty_likeYYY.

**Comment by adonath** [comment]
@registerrier On one hand I agree this leads to a simpler (and smaller) API, but on the other hand I think it introduces a lot of complexity in the code (checking of input types, checking of combinations of required arguments etc.) and leads to a much more complex documentation of the `.empty()` method.

I think "overloading" functions in Python is not very straightforward. @cdeil Did we already agree on  some standard how to handle (if at all) "overloading" of functions in gammapy?

**Comment by adonath** [comment]
@registerrier A compromise could be to allow more arguments in `SkyCube.empty_like()`. So that it either takes an existing cube or a combination of `SkyImage` and `Energy` or `EnergyBounds` objects. I.e. change the signature to `SkCube.empty_like(reference, energies=None)`. If reference is a `SkyImage` object, `energies` has to be specified in addition.

`SkyCube.reproject(reference)` already allows to specify either a `SkyImage` or `SkyCube` object as reference. Maybe this would be my favoured solution.

**Comment by registerrier** [comment]
@adonath this sounds like a good proposal.

**Comment by kosack** [comment]
I like both suggestions: 

1) to restructure the cube constructor to take an image + energy axis (it's clearer what is happening in that case)

2) add an `empty_like(x)` constructor, following the numpy convention (`zeros_like(x), ones_like(x), ...`) with either another cube or an image+energy to mimic

**Comment by kosack** [comment]
One other comment: I was confused by the name "energy", which implies a singular energy, when you mean a range. Shouldn't that be called `EnergyRange` or `EnergyBounds`, or `EnergyBins`? That would be clearer.

**Comment by cdeil** [comment]
I don't think we have an agreement or policy on when / where to put multiple simple methods, or one complex method that looks at the inputs and has complex logic about what to do.

Most of the time, I prefer simple, explicit methods. But e.g. for `SkyCoord` you can call it a dozen different ways, and it's super convenient as a user, and basically all it does is contain the logic how to dispatch to other methods. 

I do like the idea of having
```
cube = SkyCube.empty(
    image=SkyImage<some method and parameters>
    energy=Energy<some method and parameters>
)
```
as the main supported method, because it groups the arguments in a good way, and it's then also explicit what type of cube to construct, i.e. an integral one with energy bounds (like for counts), or a differential one with energy nodes (like for diffuse spectral cube model).

And if we put that, then whatever conveniences we have for SkyImage and Energy / EnergyBounds will be automatically available for SkyCube, and the SkyCube API remains small.

Now that we've heard a few opinions, anyone has time to make a pull requests? Maybe @adonath ?
In the end, I think whatever the pull request is, it'll be better than what we have now (the separate and mostly pointless helper function at the top of the cube notebook) and this could go in quickly without much further discussion.

**Comment by adonath** [comment]
Resolved by #935 . Thanks for all your suggestions!
